Removing dependency on the formatter for printing blocks.

carolahp opened the pull request:
This solution depends on the bytecode encoder and on the bytecode generator.

carolahp committed:
Removing dependency on the formatter for printing blocks.
This solution depends on the bytecode encoder and on the bytecode generator.

tesonep commented:
@carolahp There are test that are related:
- ReflectivityReificationTest>>testBlockValueAfter
- LazyClassVariableTest>>testLazyClassVariablePrint
- LazySlotTest>>testLazySlotPrint

Also, there are tests that validate if the blocks are correctly printed? 
Moreover, I want to know if there are tests to check if this works without the source code of the method.

carolahp committed:
Adding test for printing block defined in method with no source code

carolahp committed:
Replacing assert: and = by assert:equals:

carolahp committed:
Fixing definitions of blocks in  tests  for printing blocks which were failing

carolahp commented:
> @carolahp There are test that are related:
> 
> * ReflectivityReificationTest>>testBlockValueAfter
> * LazyClassVariableTest>>testLazyClassVariablePrint
> * LazySlotTest>>testLazySlotPrint
They failed because the way a block is compiled at runtime during the execution of the test is different from the compilation when the block is inside the body of a method. I corrected the way the blocks were defined in the tests. Now they pass.

> 
> Also, there are tests that validate if the blocks are correctly printed?
I found only one BlockClosureTest >> #testPrintOn 

> Moreover, I want to know if there are tests to check if this works without the source code of the method.
It works!, I added one test for this case BlockClosureTest >> #testPrintOnBlockDefinedInMethodWithoutSourceCode

MarcusDenker commented:
Nice: it uses the source code that we have already (getting the mapping where the code of the block is via the AST like the debugger). This means the pretty printer is not called anymore

MarcusDenker commented on src/Kernel/BlockClosure.class.st:
```diff
@@ -549,12 +549,7 @@ BlockClosure >> outerContext: aContext startpc: aStartpc numArgs: argCount copie
-			self class environment includesKey: #FBDDecompiler ])
-				ifTrue: [ self sourceNode formattedCode ]
-				ifFalse: [ 'aBlockClosure(no source code or decompiler available)' ])
+	aStream nextPutAll: (self method ast sourceNodeForPC: self endPC) sourceCode 
```
we can replace
` (self method ast sourceNodeForPC: self endPC) sourceCode`

with:

`self  sourceNode sourceCode`

MarcusDenker commented:
I did a test to delete the sources: debugger works fine. (in that case of course we need a pretty printer, but that is what it means to have no source: you print it from decompiled bytecode)

carolahp committed:
Adding tests for checking that the formatter is used only when the block is defined in a method without source code

carolahp committed:
Simplifying code

MarcusDenker reviewed (approved):
ok! If the tests are green we can merge

MarcusDenker commented:
failing test is related: the tests now depend on Reflectivity due to using the MetaLink... 

Given Collections do not match! additions : #(#Reflectivity) missing: #()
Stacktrace
SystemDependenciesTest(TestAsserter)>>assert:description:resumable:
SystemDependenciesTest(TestAsserter)>>assert:description:
SystemDependenciesTest(TestAsserter)>>assertCollection:hasSameElements:
SystemDependenciesTest>>testExternalSUnitKernelDependencies

carolahp committed:
Removing test that uses Metalink to avoid the dependency on Metalinks
